frontend/simple/js/transactions.js

```diff
@@ -0,0 +1,263 @@
+      // Keep track of a cursor of what step in the transaction we are operating
+      _cursor.set(this, i)
+      let currentStep = steps[ i ]
+      // Build arguments from services and scoped data
+      for (let argPair of Object.entries(currentStep.args)) {
+        args[argPair[0]] = scope[argPair[1]]
+      }
+      args.setInScope = this.setInScope.bind(this)
+      // Attempt to run step
+      try {
+        await sbp(currentStep.execute, args)
+      } catch (ex) {
+        wait.reject(new Error(`Step - ${currentStep.description || ''} failed with error, \n ${ex}`))
+        throw ex
+      }
+    }
+    wait.resolve()
+  }
+  toJSON () {
+    return {
+      steps: _steps.get(this),
+      scope: _scope.get(this),
+      cursor: _cursor.get(this),
+      isRecoverable: _isRecoverable.get(this),
+      description: _description.get(this)
+    }
+  }
+  static isTransaction (obj) {
+    if (typeof obj !== 'object') return false
+    if (!Array.isArray(obj.steps)) return false
+    if (typeof obj.scope !== 'object') return false
+    if (!Number.isInteger(obj.cursor)) return false
+    if (typeof obj.isRecoverable !== 'boolean') return false
+    if (typeof obj.description !== 'string') return false
+    return true
+  }
+  static fromJSON (json) {
+    if (!Transaction.isTransaction(json)) {
+      throw new TypeError('Invalid Transaction')
+    }
+    for (let key of Object.keys(json.scope)) {
+      if (Events.HashableEntry.isHashableEntry(json.scope[key])) {
+        json.scope[key] = Events.HashableEntry.fromJSON(json.scope[key])
+      }
+    }
+    let transaction = new Transaction(json.description, json.isRecoverable)
+    _scope.set(transaction, json.scope)
+    _cursor.set(transaction, json.cursor)
+    _steps.set(transaction, json.steps)
+    return transaction
+  }
+}
+
+export function createInternalStateTransaction (description) {
+  return new Transaction(description, false)
+}
+
+export function createExternalStateTransaction (description) {
+  return new Transaction(description, true)
+}
+// Transaction Queue
+let _transactionQueue = new WeakMap()
+let _serviceRegistry = new WeakMap()
+let _isRunning = new WeakMap()
+let _transactionID = new WeakMap()
+let _isLoaded = new WeakMap()
+let _failures = new WeakMap()
+class TransactionQueue {
+  constructor (description) {
+    _transactionQueue.set(this, [])
+    _serviceRegistry.set(this, new Map())
+    _isRunning.set(this, false)
+    _transactionID.set(this, 0)
+    _isLoaded.set(this, false)
+    _failures.set(this, [])
+  }
+  async load () {
+    if (_isLoaded.get(this)) { return }
+    try {
+      let transactionID = await transactionDB.getItem('nextID')
+      if (!transactionID) {
+        transactionID = 0
+        await transactionDB.setItem('nextID', transactionID)
+      } else {
+        await _transactionID.set(this, transactionID)
+      }
+      let failures = await transactionDB.getItem('failures') || []
+      _failures.set(this, failures)
```

Comment:
Do we need to keep separate track of "failures"? Or can we simply not pop items from the queue unless they succeed?